### PR TITLE
rgw: check for location constraint on master zonegroup

### PR DESCRIPTION
**Pull request by clwluvw**
Check the location constraint of the request when the zonegroup is master

Fixes: https://tracker.ceph.com/issues/62309

### CODE REVIEW

**Comment by cbodley** [comment]
> Limit the bucket creation with empty location constraint to master zonegroup (basically defaulting the location to the master zonegroup)

is this really what we want? i tend to think it should default to the zonegroup that issued the CreateBucket request. if the master zonegroup owns it instead, then the issuing zone/zonegroup would have to forward all object operations to the master zonegroup. in that case, the client would lose any benefits of locality with its local zone/zonegroup

cc @smanjara @mattbenjamin

**Comment by cbodley** [comment]
p.s. i reworked some of the forward_request_to_master logic in https://github.com/ceph/ceph/pull/50599. just a heads-up that we'll likely see conflicts in RGWCreateBucket::execute

**Comment by clwluvw** [comment]
@cbodley With the definition of AWS for the default region, this would make sense to have a default region and if the `LocationConstraint` was empty - pick the default one for it.
The same applies to GetBucketLocation API when the location is the default region (https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html#AmazonS3-GetBucketLocation-response-LocationConstraint).

I would say this would make sure to the clients that they are asking for the right location rather than picking the location that they are requesting (maybe by fault?).

**Comment by clwluvw** [comment]
@cbodley I couldn't follow what you meant by forwarding the request to the master. The main idea of this from my PoV was the thing I mentioned in my comment above (bucket creation and...). Are you talking about any more corners here?

**Comment by cbodley** [comment]
> With the definition of AWS for the default region, this would make sense to have a default region and if the `LocationConstraint` was empty - pick the default one for it.

@clwluvw in rgw multisite, the master zonegroup is authoritative for bucket/user metadata, but [failover](https://docs.ceph.com/en/latest/radosgw/multisite/#changing-the-metadata-master-zone) may change which zonegroup is the master. i'm not sure it makes sense to think about the master zonegroup the same way you think about a default region in aws

s3 clients/applications don't need to be aware of rgw's zonegroups at all, they just need to know the address of an http endpoint. generally they'll run some workload that creates a bucket and operates on objects in it. i think it would be unexpected if that whole workload gets redirected to a remote zonegroup instead

clients that do know about the rgw zonegroups have full control over the placement of their buckets via `LocationConstraint`. but in the absence of a location, making the local zonegroup the bucket owner seems like the obvious choice

**Comment by cbodley** [comment]
> I couldn't follow what you meant by forwarding the request to the master.

@clwluvw i was only pointing out that another PR moved around some of the same code that you're touching here

**Comment by clwluvw** [comment]
@cbodley That pretty much makes sense - Also I was worried about if we want to follow the same concept then how are we going to deal with `null` in the API I mentioned cause the SDKs are forced `eu-east-1` for that so it wouldn't make sense if the "default" is something else. I dropped the change for that.

Regarding the PR you mentioned, I guess the conflict point in the implementation I proposed is still there:
```cpp
  // validate the LocationConstraint
  if (!location_constraint.empty() && !relaxed_region_enforcement) {
    // on the master zonegroup, allow any valid api_name. otherwise it has to
    // match the bucket's zonegroup
```
Basically, I think it's better to check the LocationConstraint if the zonegroup is master as well otherwise the user might though it requested another zonegroup on the wrong HTTP endpoint and it gets 200 as a result and so the thought would be the bucket was placed in the zonegroup in bucket create request but the actual is not. What do you think?

**Comment by clwluvw** [comment]
Reproduction:
1. Have two zonegroups (assuming `A` as a master zonegroup and `B` as a secondary zonegroup)
2. Send a `CreateBucket` Request to the "zonegroup A endpoint" with the following payload:
```xml
<CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"> 
    <LocationConstraint>B</LocationConstraint> 
</CreateBucketConfiguration >
```
Results in bucket creation in zonegroup A without any error.

This is a client error as the user is asking to create the bucket in zone `B` but requesting to zonegorup `A`.
Doing that in the opposite way (request to zonegroup B with `<LocationConstraint>A</LocationConstraint>`) would result in 4xx as the endpoint would not match the requested LocationConstraint. But in the way I said - it results in 2xx and bucket creation in the master zonegroup (zone `A` - requested - but not asked) because of the `if` clause here: https://github.com/ceph/ceph/pull/52791/files#diff-319900fb50bdb02677039fb81949243b9c0e7d98690f6fddf35e6ca5d5d52b2cL3314

The PR will fix the `if` clause and check whether the zonegroup is master or not so the above example would result in 4xx as well.

**Comment by adamemerson** [comment]
jenkins test make check

**Comment by clwluvw** [comment]
jenkins test submodules

**Comment by clwluvw** [comment]
jenkins test windows

**Comment by cbodley** [comment]
> ```
> <CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"> 
>     <LocationConstraint>B</LocationConstraint> 
> </CreateBucketConfiguration >
> ```
> 
> Results in bucket creation in zonegroup A without any error.

i would hope that https://github.com/ceph/ceph/pull/59960 will cause this bucket's creation to set the correct `RGWBucketInfo::zonegroup` to zonegroup B, even though it was created on zonegroup A

**Comment by clwluvw** [comment]
> i would hope that #59960 will cause this bucket's creation to set the correct `RGWBucketInfo::zonegroup` to zonegroup B, even though it was created on zonegroup A

I don't think so as when you are master you will ignore the `location_constraint` in the body but the `rgwx-zonegroup` is playing the role now and letting the bucket be created for the correct zonegroup. so if you request directly to the master zonegroup to create a bucket for zonegroup B it won't work as it is still ignoring the body.

**Comment by cbodley** [comment]
> > i would hope that #59960 will cause this bucket's creation to set the correct `RGWBucketInfo::zonegroup` to zonegroup B, even though it was created on zonegroup A
> 
> I don't think so as when you are master you will ignore the `location_constraint` in the body but the `rgwx-zonegroup` is playing the role now and letting the bucket be created for the correct zonegroup. so if you request directly to the master zonegroup to create a bucket for zonegroup B it won't work as it is still ignoring the body.

thanks. would something like this be reasonable?
```c++
if (location_constraint) {
  bucket.zonegroup = api_name_to_zonegroup_id(location_constrant);
} else if (rgwx_zonegroup) {
  bucket.zonegroup = rgwx_zonegroup; // default to requesting zonegroup
} else {
  bucket.zonegroup = local_zonegroup_id; // default to local zonegroup
}
```

**Comment by clwluvw** [comment]
jenkins test make check arm64

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -3564,63 +3564,68 @@ void RGWCreateBucket::execute(optional_yield y)
...
+        op_ret = -ENOENT;
+        return;
+      }
+      bucket_zonegroup = &zonegroup_iter->second;
+    }
   }
 
+  // For non-master zonegroups, enforce location constraint match
+  const bool enforce_location_match = !period || !my_zonegroup.is_master;
+  if (enforce_location_match && !my_zonegroup.equals(bucket_zonegroup->get_id())) {
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:3608]: is there a reason that other zonegroups shouldn't also support creation of remote buckets?

in aws, it looks like you can send a [CreateBucket](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html) request to any region, naming any other valid region in the LocationConstraint

Comment (clwluvw) [reply src/rgw/rgw_op.cc:3608]: No you will get `IllegalLocationConstraintException` is response. and tbh I would still be voting for master to also not create bucket for other zonegroups when the request is initiated by the user - not forwarded.
This would make some races easier to solve deleting the condition that we need to cover.

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: > +1

:face_holding_back_tears: 


I just drafted a test case for this as well, will try to verify on my local asap.

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: The tests have passed on my local, but we will need https://github.com/ceph/ceph/pull/59960.

Comment (cbodley) [reply src/rgw/rgw_op.cc]: > but we will need #59960

working on it

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: > > but we will need #59960
> 
> working on it

If you are running QA on that, perhaps it might be worth including this as well and dropping the test case there, as this case here is more accurate and covers both (creating a bucket and ensuring the bucket has been created for the corrected zonegroup).

Comment (cbodley) [reply src/rgw/rgw_op.cc]: i cherry-picked the commit from this PR on top of the qa branch for https://github.com/ceph/ceph/pull/59960 and removed its test case. https://pulpito.ceph.com/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/ shows a similar failure for the two-zonegroup job (the two-zone job hasn't run yet):
```
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:======================================================================
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:ERROR: rgw_multi.tests.test_bucket_create_location_constraint
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_21fbe25bb2cf151e2eb892c1d86880b6eb21b2b5/qa/../src/test/rgw/rgw_multi/tests.py", line 3651, in test_bucket_create_location_constraint
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    z.s3_client.create_bucket(Bucket=bucket_name, CreateBucketConfiguration={'LocationConstraint': zg.name})
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/botocore/client.py", line 553, in _api_call
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    return self._make_api_call(operation_name, kwargs)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/botocore/client.py", line 1009, in _make_api_call
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    raise error_class(parsed_response, operation_name)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:botocore.exceptions.ClientError: An error occurred (InvalidLocationConstraint) when calling the CreateBucket operation: The specified location-constraint is not valid
```

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: By any chance, do we have access to rgw logs?

Comment (cbodley) [reply src/rgw/rgw_op.cc]: zone a1: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi043/log/rgw.c1.client.0.log.gz
zone a2: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi104/log/rgw.c1.client.1.log.gz
zone b1: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi160/log/rgw.c2.client.0.log.gz
zone b2: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi187/log/rgw.c2.client.1.log.gz

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: I'm thinking maybe there is something fishy about the `zonegorups_by_api` var.
Currently, there are only two places the variable is getting filled, `RGWPeriodMap::update()` and `RGWPeriodMap::decode()`. I guess decode is irrelevant here so we should see where we have `RGWPeriodMap::update()` called.
I guess it should be `RGWPeriod::update()` but I can't find any logs saying `RGWPeriod::update() realm ...` due to the second line of the function logging with no condition.
the other possibility is `update_period()` which is only called by rgw-admin where perhaps on `period update --commit` commands it should construct that for itself but still on the master zonegorup and master zone `a1` (which we are sure the zonegroup `a` has been created) it cannot find itself in the period map
```
2024-12-19T18:26:36.873+0000 7f1be8535640  0 req 13354008642218923451 0.000000000s s3:create_bucket location constraint (a) can't be found.
```

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: ok this is actually weird - api_name is empty.
```
2024-12-19T18:15:53.211 INFO:teuthology.orchestra.run.smithi043.stdout:{"id":"8987fccd-6980-4682-9cce-cf39a4aa6b34","epoch":12,"predecessor_uuid":"5276a8ed-77a4-4898-9669-84e295758977","sync_status":[],"period_map":{"id":"8987fccd-6980-4682-9cce-cf39a4aa6b34","zonegroups":[{"id":"a","name":"a","api_name":"","is_master":true,"endpoints":["http://smithi043.front.sepia.ceph.com:8000"],"hostnames":[],"hostnames_s3website":[],"master_zone":"6a93f239-7b93-4dfe-94b1-baf12183eb58","zones":[{"id":"250dd5e0-4049-46ca-bf5b-e7e54b3b29b4","name":"a2","endpoints":["http://smithi104.front.sepia.ceph.com:8001"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]},{"id":"6a93f239-7b93-4dfe-94b1-baf12183eb58","name":"a1","endpoints":["http://smithi043.front.sepia.ceph.com:8000"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]}],"placement_targets":[{"name":"default-placement","tags":[],"storage_classes":["STANDARD"]}],"default_placement":"default-placement","realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","sync_policy":{"groups":[]},"enabled_features":["notification_v2","resharding"]},{"id":"b","name":"b","api_name":"","is_master":false,"endpoints":["http://smithi160.front.sepia.ceph.com:8000"],"hostnames":[],"hostnames_s3website":[],"master_zone":"b3be8bac-85cd-4275-8c15-00f60f318a67","zones":[{"id":"b3be8bac-85cd-4275-8c15-00f60f318a67","name":"b1","endpoints":["http://smithi160.front.sepia.ceph.com:8000"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]},{"id":"e868f7a7-a7b3-4202-8e6f-86642035ca92","name":"b2","endpoints":["http://smithi187.front.sepia.ceph.com:8001"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]}],"placement_targets":[{"name":"default-placement","tags":[],"storage_classes":["STANDARD"]}],"default_placement":"default-placement","realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","sync_policy":{"groups":[]},"enabled_features":["notification_v2","resharding"]}],"short_zone_ids":[{"key":"250dd5e0-4049-46ca-bf5b-e7e54b3b29b4","val":104308217},{"key":"6a93f239-7b93-4dfe-94b1-baf12183eb58","val":2854171449},{"key":"b3be8bac-85cd-4275-8c15-00f60f318a67","val":1981456006},{"key":"e868f7a7-a7b3-4202-8e6f-86642035ca92","val":3492425437}]},"master_zonegroup":"a","master_zone":"6a93f239-7b93-4dfe-94b1-baf12183eb58","period_config":{"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1},"user_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1},"user_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false},"bucket_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false},"anonymous_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false}},"realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","realm_epoch":2}
```

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: hmm.. I'm not sure why we create zonegroups via `zonegroup set` but that might explain why api_name is created empty https://github.com/ceph/ceph/blob/862ed6e9cd8cece7edc6d7d45505b5a4b4f00e26/qa/tasks/rgw_multisite.py#L357-L369

Comment (clwluvw) [reply src/rgw/rgw_op.cc]: @cbodley I pushed a fix (c21b5f7e0b1a48e7a4b4e21d1d357e8f2f26a1ac), can you try again?

Comment (cbodley) [reply src/rgw/rgw_op.cc]: passed multisite jobs in https://pulpito.ceph.com/cbodley-2025-01-07_21:31:35-rgw:multisite-wip-69281-distro-default-smithi/

> rgw_multi.tests.test_bucket_create_location_constraint ... ok

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2025-01-08_19:44:28-rgw-wip-69281-distro-default-smithi/ (batched with https://github.com/ceph/ceph/pull/59960)
